### PR TITLE
Author new linter rule for checking if the request body is present for Put and Patch requests 

### PR DESCRIPTION
**Pull request by rkmanda**
This is to automate 

RPC-Put-V1-28 
Put must have a request body defined

RPC-Patch-V1-12
Patch must have a request body

### CODE REVIEW

File: packages/rulesets/generated/spectral/az-arm.js [hunk 1]
```diff
@@ -2690,6 +2690,27 @@ const validatePatchBodyParamProperties = createRulesetFunction({
     return errors;
 });
 
+const requestBodyMustExistForPutPatch = (putPatchOperationParameters, _opts, ctx) => {
+    const errors = [];
+    const path = ctx.path;
+    const error = `The put or patch operation does not have a request body defined. This is not allowed. Please specify a request body for this operation.`;
+    const bodyParam = findBodyParam(putPatchOperationParameters);
+    if (bodyParam == undefined || bodyParam["schema"] == undefined) {
```
Comment (bdefoy) [inline packages/rulesets/generated/spectral/az-arm.js:2698]: Might want to check for empty objects here as well:
```suggestion
    if (bodyParam == undefined || bodyParam["schema"] == undefined || typeof bodyParam["schema"] !== 'object' || _.isEmpty(bodyParam["schema"])) {
```

File: packages/rulesets/generated/spectral/az-arm.js [hunk 2]
```diff
@@ -2690,6 +2690,27 @@ const validatePatchBodyParamProperties = createRulesetFunction({
...
+        errors.push({
+            message: error,
+            path: path,
+        });
+    }
+    return errors;
+};
+function findBodyParam(params) {
```
Comment (bdefoy) [inline packages/rulesets/generated/spectral/az-arm.js:2706]: You can probably put this in utils.ts

File: packages/rulesets/src/spectral/az-arm.ts [hunk 3]
```diff
@@ -559,6 +560,19 @@ const ruleset: any = {
       },
     },
 
+    // RPC Code: RPC-Put-V1-28, RPC-Patch-V1-12
+    RequestBodyMustExistForPutPatch: {
+      description: "Every Put operation must have a request body",
```
Comment (bdefoy) [inline packages/rulesets/src/spectral/az-arm.ts:565]:
```suggestion
      description: "Every Put and Patch operation must have a request body",
```

File: packages/rulesets/src/spectral/test/request-body-must-exist-for-put-patch.test.ts
Comment (bdefoy) [inline packages/rulesets/src/spectral/test/request-body-must-exist-for-put-patch.test.ts:1]: Maybe add a test for body parameter with an empty schema: `{}`

File: docs/rules.md [hunk 4]
```diff
@@ -987,12 +980,30 @@ For any PUT, response code should be 201 if resource was newly created and 200 i
...
+
+Synchronous and long-running PUT operations must have responses with 200, 201 and default return codes. They also must not have other response codes.
+
+Please refer to [put-response-codes.md](./put-response-codes.md) for details.
+
+### PutResponseSchemaDescription
+
+For any PUT, response code should be 201 if resource was newly created and 200 if updated.
+
+Please refer to [put-response-schema-description.md](./put-response-schema-description.md) for details.
```
Comment (AkhilaIlla) [inline docs/rules.md:993]: I think we deleted this rule & shouldnt be in here?

File: docs/request-body-must-exist-for-put-patch.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+ARM OpenAPI(swagger) specs
+
+
+## Related ARM Guideline Code
+
+- RPC-Put-V1-28, RPC-Patch-V1-12
+
+## Output Message
+
+A Put or Patch request must always have a request body defined.
```
Comment (AkhilaIlla) [inline docs/request-body-must-exist-for-put-patch.md:18]: >Put or Patch [](http://example.com/codeflow?start=2&length=12)

Is it PUT or PATCH?
or PUT and PATCH

File: docs/request-body-must-exist-for-put-patch.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+A Put or Patch request must always have a request body defined. 
+
+## Description
+This rule applies for tracked as well as proxy resources. This applies even in the case where there are no properties defined in the properties bag. 
+
+
+## Why the rule is important
+
+Making the request body mandatory for the Put or Patch request is to ensure that the output of the Get request may be passed in to the Put request and that can be achieved only if the Put has a request body defined. For Patch, it does not make sense to patch an empty payload.
```
Comment (AkhilaIlla) [inline docs/request-body-must-exist-for-put-patch.md:26]: > Put or Patc [](http://example.com/codeflow?start=41&length=12)

Is it PUT or PATCH?
or PUT and PATCH

File: packages/rulesets/src/spectral/az-arm.ts [hunk 7]
```diff
@@ -559,6 +560,19 @@ const ruleset: any = {
...
+      message: "{{error}}",
+      severity: "error",
+      resolved: true,
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/az-arm.ts:568]: >    resolved: true, [](http://example.com/codeflow?start=2&length=19)

set stagingonly

File: packages/rulesets/generated/spectral/az-arm.js (hunk 1 above)
Comment (rkmanda) [reply packages/rulesets/generated/spectral/az-arm.js:2698]: This condition - typeof bodyParam["schema"] !== 'object'  would not be valid as the value will be a string. Added the other condition for now. We can add deeper validation if required later.

File: docs/request-body-must-exist-for-put-patch.md (hunk 6 above)
Comment (rkmanda) [reply docs/request-body-must-exist-for-put-patch.md:26]: This sentence should only have Put specified. Will correct it

File: docs/rules.md (hunk 4 above)
Comment (rkmanda) [reply docs/rules.md:993]: yeah not sure what got it in here, will check

Comment (rkmanda) [reply docs/rules.md]: fixed after merging from main

**Review by AkhilaIlla — APPROVED** [review]

File: packages/rulesets/generated/spectral/az-arm.js (hunk 1 above)
Comment (bdefoy) [reply packages/rulesets/generated/spectral/az-arm.js]: I believe it would be valid since you have `resolved: true` set for this rule. When that is set it should replace all of the `$ref` items with their actual values. However, we might already have some mechanism that checks for empty objects anyway.

File: packages/rulesets/generated/spectral/az-arm.js (hunk 2 above)
Comment (rkmanda) [reply packages/rulesets/generated/spectral/az-arm.js]: It is in utils. This is the generated code. @bdefoy
